src/Microsoft.ML.LightGbm/LightGbmArguments.cs

```diff
@@ -1,678 +1,225 @@
+                    "half of the data instances to grow trees and this will prevent overfitting. Range: (0,1].")]
+            [TlcModule.Range(Inf = 0.0, Max = 1.0)]
+            public double SubsampleFraction = 1;
+
+            [Argument(ArgumentType.AtMostOnce,
+                HelpText = "Subsample ratio of columns when constructing each tree. Range: (0,1].",
+                ShortName = "ff")]
+            [TlcModule.Range(Inf = 0.0, Max = 1.0)]
+            public double FeatureFraction = 1;
+
+            [Argument(ArgumentType.AtMostOnce,
+                HelpText = "L2 regularization term on weights, increasing this value will make model more conservative.",
+                ShortName = "l2")]
+            [TlcModule.Range(Min = 0.0)]
+            [TGUI(Label = "Lambda(L2)", SuggestedSweeps = "0,0.5,1")]
+            [TlcModule.SweepableDiscreteParam("RegLambda", new object[] { 0f, 0.5f, 1f })]
+            public double L2Regularization = 0.01;
+
+            [Argument(ArgumentType.AtMostOnce,
+                HelpText = "L1 regularization term on weights, increase this value will make model more conservative.",
+                ShortName = "l1")]
+            [TlcModule.Range(Min = 0.0)]
+            [TGUI(Label = "Alpha(L1)", SuggestedSweeps = "0,0.5,1")]
+            [TlcModule.SweepableDiscreteParam("RegAlpha", new object[] { 0f, 0.5f, 1f })]
+            public double L1Regularization = 0;
+
+            [Argument(ArgumentType.AtMostOnce,
+                HelpText = "Control the balance of positive and negative weights, useful for unbalanced classes." +
+                " A typical value to consider: sum(negative cases) / sum(positive cases).",
+                ShortName = "ScalePosWeight")]
+            public double WeightOfPositiveExamples = 1;
+
+            BoosterParameterBase IComponentFactory<BoosterParameterBase>.CreateComponent(IHostEnvironment env)
             {
-                BoosterParameterOptions = options;
+                return BuildOptions();
             }
 
-            /// <summary>
-            /// Update the parameters by specific Booster, will update parameters into "res" directly.
-            /// </summary>
-            internal virtual void UpdateParameters(Dictionary<string, object> res)
+            BoosterParameterBase IBoosterParameterFactory.CreateComponent(IHostEnvironment env)
             {
-                FieldInfo[] fields = BoosterParameterOptions.GetType().GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
-                foreach (var field in fields)
-                {
-                    var attribute = field.GetCustomAttribute<ArgumentAttribute>(false);
-
-                    if (attribute == null)
-                        continue;
-
-                    res[GetOptionName(field.Name)] = field.GetValue(BoosterParameterOptions);
-                }
+                return BuildOptions();
             }
```

Comment:
Not your problem. C# doesn't work well here, so we must have two (conceptually) identical implementations..

Reply:
thanks for mentioning this, I will keep this active for other reviewers.

---
In reply to: [265815338](https://github.com/dotnet/machinelearning/pull/2948#discussion_r265815338) [](ancestors = 265815338)